activerecord/lib/arel/visitors/postgresql.rb

```diff
@@ -36,6 +36,45 @@ def visit_Arel_Nodes_NotRegexp(o, collector)
           infix_value o, collector, op
         end
 
+        def visit_Arel_Nodes_HomogeneousIn(o, collector)
```

Comment:
Instead of "hi-jacking" `HomogeneousIn` node, it should be also possible to use custom node and check for attribute type in ArrayHandler.

https://github.com/rails/rails/blob/b71a9ccce04ac08e159d4a21de91a8d76f13d8d0/activerecord/lib/active_record/relation/predicate_builder/array_handler.rb#L19-L24